File: javascript/bob/bob.js
class Bob {
  hey(message) {
    if(this.shouting(message) && !this.numbers(message)) {
      return "Whoa, chill out!";
    }
    else if(this.question(message)) {
      return "Sure.";
    }
    else if(!message.trim()) {
      return "Fine. Be that way!";
    }
    else {
      return "Whatever.";
    }
  }

  shouting(message) {
    let uppercase = message.toUpperCase();
    return uppercase == message;
  }

  question(message) { return message.endsWith("?"); }

  numbers(message) { return !/[a-zA-Z]/.test(message); }
}

export default Bob;

File: elixir/bob/bob.exs
defmodule Bob do

  def hey(input) do
    cond do
      question?(input) ->
        "Sure."
      silence?(input) ->
        "Fine. Be that way!"
      shouting?(input) && !numbers?(input) ->
        "Whoa, chill out!"
      true ->
        "Whatever."
    end
  end

  defp shouting?(input) do
    String.upcase(input) == input
  end

  defp question?(input) do
    String.ends_with?(input, "?")
  end

  defp silence?(input) do
    String.strip(input) == ""
  end

  defp numbers?(input) do
    String.match?(input, ~r/^[0-9\s,\?!]+$/)
  end

end
